Update annotation import

hudson-newey opened the pull request:
# Update annotation import

Following breaking changes to the annotation import API, I have restructured the annotation import API and UI to reflect these changes.

## Changes

- Adds `AnnotationImportFile` models & services
- Adds error card component
- Adds `bawVirtualDatatablePagination` directive so that we can use table pagination without an API backing
- Adds new `typeaheadCallback` method to the `tags.service`
- Adds handling of edge case where the typeahead input doesn't have a callback instantiated
- Updates `AnnotationImport` model
- Updates `ImportedAudioEvent` model
- `AbstractModel` `getJsonAttributes(opts: Xor<{ create: boolean} | { update: boolean }>)` has been updated to `getJsonAttributesForCreate` and `getJsonAttributesForUpdate`
- The `baw-api` service `create` and `update` methods now accept a `params` object for unscoped form data
- Updates `BawApiError`s to provide the responses `data`
- Removes `AnnotationImportFileRead` and `AnnotationImportFileWrite` models (models returned for reading / writing used to be different. This was fixed in the API refactor)

## Issues

Fixes: #2200

## Potential future changes

- When clicking on the "Add annotations" button, it should take you to a combined page to create a new annotation import and add files
- Users might want to edit their annotation before they commit them. (have input boxes so that they can change the values)

## Visual Changes

![image](https://github.com/user-attachments/assets/2afacde5-b8c2-4617-ba3f-154fc5a40fbb)

_Annotation import details page (events)_

![image](https://github.com/user-attachments/assets/5c864fb7-56ed-498c-820d-c2d3067ce745)

_Annotation import details page (files)_

![image](https://github.com/user-attachments/assets/6a8d5179-75c0-4f7c-a6b8-fda3e26c06b1)

_Add new annotations page (with no uploaded files)_

![image](https://github.com/user-attachments/assets/70895e33-698a-4c21-b57c-e16e88037b82)

_Add new annotations page (with errors)_

(note: with the new columns, the table is now horizontally scrollable)

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Update models

hudson-newey committed:
Update services

hudson-newey committed:
Use new services in details page

hudson-newey committed:
Move upload logic to add-annotation.component

hudson-newey committed:
Conditional PUT or POST requests for create api

hudson-newey committed:
Bifurcate attribute serializers

hudson-newey committed:
Add preview to import

hudson-newey committed:
Use ngx-datatable for indentified annotations

hudson-newey committed:
Successful dry run

hudson-newey committed:
Add error card component

hudson-newey committed:
Get tests working and add error card tests

hudson-newey committed:
Correctly await all requests succeding in commit

hudson-newey committed:
Various improvements

hudson-newey committed:
Improve file table in import details view

hudson-newey committed:
Add form input guard

hudson-newey committed:
Fix edit tests

hudson-newey committed:
Add annotation smoke tests

hudson-newey committed:
Add view button to import event details table

hudson-newey committed:
import detail component smoke tests

hudson-newey committed:
Add import detail specs

hudson-newey committed:
Remove ImportGroup interface

hudson-newey committed:
Update abstract model spec

hudson-newey committed:
Add baw-api.service tests for data errors

hudson-newey committed:
Fix baw-api-error spec

hudson-newey committed:
Fix error card tests

hudson-newey committed:
Add missing service for tags typeahead

hudson-newey committed:
Fix loading state when no files were uploaded

hudson-newey committed:
Add support for file wide errors

hudson-newey committed:
Fix file error handling

hudson-newey committed:
Backfill add annotations page

hudson-newey committed:
Staging

hudson-newey committed:
Fix csv files being sent as excel types

hudson-newey committed:
Temp fix for pagination directive with no rows

hudson-newey committed:
Add basic tests for type conversion and dry runs

hudson-newey committed:
Add navigation warning tests

hudson-newey committed:
Add commit run tests

hudson-newey committed:
Add file error tests

hudson-newey committed:
Complete dry run tests

hudson-newey committed:
Show file name when file errors

hudson-newey committed:
'gracefully' handle internal server errors

hudson-newey committed:
Add virtualDatatablePagination directive

hudson-newey committed:
backfill virtual pagination directive tests

hudson-newey committed:
Better organise datatable directives

hudson-newey committed:
Fix virtual datatable pagination file casing

hudson-newey committed:
Finish virtual pagination tests

hudson-newey committed:
Use ClassBuilder constructor for BawApiError data

hudson-newey committed:
Fix error card tests

hudson-newey committed:
Fix linting errors

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.html:
```diff
@@ -0,0 +1,228 @@
+            <a [bawUrl]="value.audioRecording.viewUrl">
+              {{ value.audioRecording.id }}
+            </a>
+          </ng-template> -->
```
Useless?

hudson-newey replied:
Resolved by implementing router links for audio recording column

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,360 @@
+  protected valid: boolean = false;
+  protected uploading: boolean = false;
+  protected importFiles: File[] = [];
+  protected additionalTagIds: Id[] = [];
```
Smells: additional tags are per file. The relationship seems mismatched?

atruskie replied:
oh I think I get it now. it is one to one, but multiple files can be imported from the file picker?

hudson-newey replied:
Following a later comment, and an offline discussion, we've decided that you should be able to add additional tags to each individual file, along with being able to add additional tags to every file in the current queue

hudson-newey committed:
Handle cases where the user doesn't have recording access

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.html:
```diff
@@ -0,0 +1,237 @@
+  </p>
+
+  <div>
+    <ngx-datatable
```
can this data table show events from more than one file? If so, we need a file column.

Probably need an index column too.

Maybe this could be combined: 

| ID (tooltip: "file index and even index in each file") |
|----|
| 0:0 |
| 1:0 |
| 1:1234|

hudson-newey replied:
Yes it can, I like the idea

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+   * table.
+   * E.g. duplicate file uploads, names, etc...
+   */
+  protected importErrors: BawErrorData[] = [];
```
i'm again wondering what the arity is here.

if these are structured how i'm assuming, perhaps it should be an array of objects, rather than three arrays

hudson-newey replied:
![image](https://github.com/user-attachments/assets/a2d84098-9041-42ec-9ba8-4ce90b5decf5)

The page component now uses an array of `QueuedFile[]` to remove these arrays

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    // we should never hit this condition, but if we do, I return null so that
+    // we don't end up in an unknown state
```
nit: concise comments are less load

```suggestion
    // should never hit, but be safe
```

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+        contains<Tag, "text">("text", text),
+        notIn<Tag>("text", activeItems)
+      ),
+    });
```
feels like we use this a lot, and thus if it is here, it isn't DRY.... factor out?

hudson-newey replied:
I've added a `typeaheadCallback` method to the tags.service which we can use in any component

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+    });
+
+  // since the typeahead input returns an array of models, but the api wants the associated tags as an array of id's
+  // we use this helper function to convert the array of models to an array of id's that can be sent in the api request
```
Perhaps the filter should accept an array of tags or ids? We use that convention elsewhere

```suggestion
  // typeahead returns models, we need ids for filter
```

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+    // an error.
+    // to fix this, we will change the file type to the correct type using the
+    // file extension.
+    const extensionMappings = new Map<string, string>([["csv", "text/csv"]]);
```
don't redeclare this map in the handler each time. Make it a real constant

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+  }
+
+  protected hasRecordingErrors(model: ImportedAudioEvent): boolean {
+    return model.errors.some((error) => "audioRecordingId" in error);
```
is this a constrained string checked by TS? otherwise magic string

hudson-newey replied:
This is 50% type checked (good catch, it can be improved)

If I change the function signature to something like (this code below doesn't actually work, but it just some lookalike pseudo code)

```ts
// (could alternatively use `infer`)
protected hasRecordingError<T extends ImportedAudioEvent>(model: T): "audioRecordingId" in T["errors"]
```

Then using this function on a model would actually restrict the model type.

---

I can also restrict the `error` properties type further by improving the typing of this interface

https://github.com/QutEcoacoustics/workbench-client/blob/master/src/app/models/AudioEventImport/ImportedAudioEvent.ts#L11-L13

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+      return;
+    }
+
+    // creates a lock so that no more files can be added to the upload queue while the upload is in progress
```
this comment belongs on the field declaration

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    // creates a lock so that no more files can be added to the upload queue while the upload is in progress
+    this.uploading = true;
```
we use this pattern a lot. Future idea a `Flag` class. Methods like `this.uploading.lock()`, `this.uploading.unlock()`, `this.uploading.lockFor( () => { } )`. It can also have helpers for locking and unlocking with promises/observables.

hudson-newey replied:
Related issue

https://github.com/QutEcoacoustics/workbench-client/issues/2216

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+            })
+          );
+        },
+      });
```
what happens on failure?

hudson-newey replied:
Failure during a wet run should never happen (unless the wet run fails and the dry run passes).

I can think of one time when this might happen:
If the users access to a project is revoked between the dry run finishing and committing the annotations.

In this case, the user would go through the normal error handling, with a toast message.

To handle this edge case, I can display the errors in the preview table (similar to a dry run)

atruskie replied:
yeah you should add error handling code.

you don't know what has changed since the last dry run

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+        //
+        // TODO: this should be properly fixed by modifying the typing of the
+        // BawApiError class
+        return of(error.data as AudioEventImportFile);
```
seems to be worth fixing

hudson-newey replied:
![image](https://github.com/user-attachments/assets/bcd345e6-b846-458b-b4fd-520c52357eba)

After reviewing the typing of `error.data`, this comment is incorrect, the problem is that `error.data` can be an array of errors, or a single array.

Because we expect a single error object for this type of response I have added a condition (that should never hit) to remove the `array` type from details

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+      }
+    }
+
+    return newErrors;
```
seems like some of this should be part of the service/error models

hudson-newey replied:
I'm a bit hesitant to mutate the error message on the base response model, so I wouldn't want to modify/mutate the error message that is returned from the `dryCreate` method.

---

What I'm thinking from your proposal is:

A new method on the `audio-event-import-file.service` (e.g. `humanize-errors` that takes a `File` object, and a `BawApiError` model.

It will change all of the `file` keys on the `BawApiError` details to use the `File.name` instead

hudson-newey replied:
Discussed offline, we shouldn't be modifying the BawApiError message at all, this should be a purely templating change.

hudson-newey replied:
![image](https://github.com/user-attachments/assets/3b5092e1-6871-4737-b895-1742fc497fc7)

atruskie commented on src/app/components/import-annotations/add-annotations/add-annotations.component.ts:
```diff
@@ -0,0 +1,364 @@
+    const newErrors: BawErrorData[] = Array.isArray(error.info)
+      ? Object.assign({}, error.info)
+      : [Object.assign({}, error.info)];
+
```
why mutate at all?

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -78,7 +90,91 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+
+  public getUnscopedAttributes(): any[] {
+    return [];
+  }
```
not keen.  We don't need such broad functionality.

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -78,7 +90,91 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+
+  public getJsonAttributes(): Partial<this> {
+    return this.jsonAttributes();
+  }
```
are these actually used anywhere?

hudson-newey replied:
Only in tests... I will remove them

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -78,7 +90,91 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+
+  public getFormDataOnlyAttributes(): FormData {
+    return this.formDataOnlyAttributes();
+  }
```
concise docs needed for all of these

```
/* { @bawPersistAttr({create: true} x = 1, @bawPersistAttr({supportedFormats: ['form-data']}) y = 2
 *  getFormDataOnlyAttributesForCreate() => { y = 2 }
 */  
```

atruskie commented on src/app/models/AttributeDecorators.ts:
```diff
@@ -284,6 +284,10 @@ export function bawBytes<Model>(opts?: BawDecoratorOptions<Model>) {
+export function unscopedAttribute<Model>(opts?: BawDecoratorOptions<Model>) {
+  return createDecorator<Model>(opts, () => {});
+}
+
```
👎

atruskie commented on src/app/models/AudioEvent.ts:
```diff
@@ -66,7 +73,9 @@ export class AudioEvent
   public readonly updatedAt?: DateTimeTimezone;
   @bawDateTime()
   public readonly deletedAt?: DateTimeTimezone;
+  @bawPersistAttr()
```
hmm. provenanceId isn't writable. Why have a persistAttr?

atruskie commented on src/app/models/AudioEvent.ts:
```diff
@@ -56,6 +61,8 @@ export class AudioEvent
   @bawPersistAttr()
   public readonly isReference?: boolean;
+  @bawPersistAttr()
+  public readonly score?: number;
```
score isn't writable

atruskie commented on src/app/models/AudioEvent.ts:
```diff
@@ -47,6 +50,8 @@ export class AudioEvent
   @bawPersistAttr()
   public readonly audioRecordingId?: Id;
   @bawPersistAttr()
+  public readonly channel?: number;
```
channel isn't writable - but it should be. File a bug and comment out

hudson-newey replied:
Related issue: https://github.com/QutEcoacoustics/baw-server/issues/717

atruskie commented on src/app/models/AudioEventImportFile.ts:
```diff
@@ -0,0 +1,68 @@
+  public readonly name?: string;
+  @bawPersistAttr()
+  public readonly fileHash?: string;
+  @bawPersistAttr()
```
baw persist attrs

atruskie commented on src/app/models/AudioEventImportFile.ts:
```diff
@@ -0,0 +1,68 @@
+
+  // form data fields
+  @bawPersistAttr({ supportedFormats: ["formData"], create: true })
+  public readonly analysisJobsItemId?: Id;
```
ajii is not writable in the API

atruskie commented on src/app/models/AudioEventImportFile.ts:
```diff
@@ -0,0 +1,68 @@
+  public readonly path?: FilePath;
+  @bawDateTime()
+  public readonly createdAt?: DateTimeTimezone;
+  public readonly committed?: boolean;
```
maybe add docs: this field is only present immediately after creation (dry or not)

atruskie commented on src/app/models/AudioEventImportFile.ts:
```diff
@@ -0,0 +1,68 @@
+  public readonly createdAt?: DateTimeTimezone;
+  public readonly committed?: boolean;
+  @bawSubModelCollection<AudioEventImportFile, ImportedAudioEvent>(ImportedAudioEvent)
+  public readonly importedEvents?: ImportedAudioEvent[];
```
maybe add docs: this field is only present immediately after creation (dry or not)

atruskie commented on src/app/services/baw-api/api-common.ts:
```diff
@@ -71,6 +71,7 @@ export type Filter = "filter";
 export const emptyParam: Empty = "";
 export const newParam: New = "new";
 export const filterParam: Filter = "filter";
+export const commitParam = "?commit";
```
this is not a common pattern, this constant should probably not be in common. Move to AEIF service

atruskie commented on src/app/services/baw-api/audio-event-import-file/audio-event-import-file.service.ts:
```diff
@@ -0,0 +1,95 @@
+    );
+  }
+
+  public dryCreate(
```
add doc: "like create but does not result in a saved record. Specifically for testing if a payload passes validation"

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -42,6 +44,9 @@ export interface BawServiceOptions {
   /** If set, requests will include the users authentication token and cookies */
   withCredentials?: boolean;
 
+  /** Unscoped form data to be sent with the request */
```
```suggestion
  /** Additional parameters to merged in with the model payload to be sent with the request */
```

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -386,12 +460,16 @@ export class BawApiService<
+    let formData = model.getFormDataOnlyAttributesForUpdate();
+    if (options.params) {
+      formData = this.addUnscopedParams(formData, options.params);
+    }
```
should probably also handle a json case (in which case just merge with body, or QSP is acceptable too)

atruskie commented on src/app/test/fakes/AudioEventImport.ts:
```diff
@@ -1,49 +1,19 @@
+    name: modelData.param(),
+    importedEvents: modelData.randomArray(5, 20, () =>
+      generateImportedAudioEvent()
+    ),
```
not a prop on this model

atruskie commented on src/app/test/fakes/AudioEventImportFile.ts:
```diff
@@ -0,0 +1,24 @@
+    ),
+    analysisJobsItemId: modelData.id(),
+    audioEventImportId: modelData.id(),
+    file: modelData.file(),
```
https://github.com/QutEcoacoustics/baw-server/blob/f18d7bf248aae05fd22e62614db33d244ea930a6/app/models/audio_event_import_file.rb#L278-L287

hudson-newey committed:
Review feedback

hudson-newey committed:
Staging review feedback

hudson-newey committed:
Style changes

hudson-newey committed:
Add file list

hudson-newey committed:
Handle json payloads for unscoped data

hudson-newey committed:
Add comments for new AbstractModel methods

hudson-newey committed:
Fix serialization for FormData item arrays

hudson-newey committed:
Remove re-declaration of extensionMappings

hudson-newey committed:
Refactor out tag typeahead callback

hudson-newey committed:
File errors are now templated

[no ci]

hudson-newey committed:
Improve mock models

hudson-newey committed:
Handle errors when performing wet run

hudson-newey committed:
Style cleanup

hudson-newey committed:
Correctly handle errors

hudson-newey committed:
Fix importing multiple files

hudson-newey committed:
Improve code quality

hudson-newey committed:
fix duplicate files not showing errors

hudson-newey committed:
Reactively update file input when removing files

hudson-newey committed:
Update tests

hudson-newey committed:
fix ability to add additional tags per file

This was previously possible, but did not include a bunch of edge cases.
E.g. uploading a duplicate file

hudson-newey committed:
Improve navigation warning

hudson-newey committed:
fix test bed

hudson-newey committed:
Correctly update file input text when adding files

hudson-newey committed:
fix misaligned tooltip

hudson-newey committed:
Staging

hudson-newey committed:
Refactor table row view model

hudson-newey committed:
Fix result table

hudson-newey committed:
fix missing predicates in template

hudson-newey committed:
fix extra tags

hudson-newey committed:
fix falsy values showing as "missing" in dataset

hudson-newey committed:
Fix 0 score showing as missing

hudson-newey committed:
offline review feedback

hudson-newey committed:
Emit audio recording ids in error states

hudson-newey committed:
All previous tests passing

hudson-newey committed:
Add navigation warning tests

hudson-newey committed:
Add tests for removing files

hudson-newey committed:
Fully working with tests

hudson-newey committed:
Fix both dry and commit requests firing on upload

This was a bug from using iif in the previous commit

hudson-newey committed:
Consistent formatting location for pagination.directive

hudson-newey committed:
Populate import detail test table

hudson-newey committed:
Tests for annotation detail event table

hudson-newey committed:
Fix linting issues

atruskie commented on src/app/services/baw-api/audio-event-import-file/audio-event-import-file.service.spec.ts:
no tests?

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -316,31 +383,56 @@ export class BawApiService<
+        });
+
+        path = url.href;
+      }
```
why not just append unscoped vars to the jsonData body?

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -386,12 +480,32 @@ export class BawApiService<
+          url.searchParams.append(key, value);
+        });
+
+        path = url.href;
```
as above

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -691,6 +807,17 @@ export class BawApiService<
+  ): FormData {
+    for (const [key, value] of Object.entries(params)) {
+      data.append(key, value);
+    }
```
seems we just special cased adding arrays to formdata, but we haven't done it here. is that an issue?

hudson-newey replied:
Really good catch (thank you).

TypeScript should have caught this so I think an underlying issue here is that the `params` property on the options is a `Record<any, any>` so I didn't get any TypeScript errors to warn me about this.

![image](https://github.com/user-attachments/assets/9596dc3e-d1a2-44f0-a9ca-32bc7fd469d1)

hudson-newey committed:
Add import-file.service tests

hudson-newey committed:
Fix failing tests and remove "fit"
